### PR TITLE
feat: allow multiple port and protocols for gateway

### PR DESCRIPTION
**Pull request by alaeddine-13**
This PR adds support for multiple ports and protocols for the gateway runtime.

- [x] Gateway parser can receive list or single value
- [x] Flow `__init__` and `config_gateway` can receive multiple or single value. Flow.port and Flow.protocol return single or multiple values
- [x] GatewayRuntime expects list of values
- [x] BaseGateway can receive runtime args
- [ ] (optional) move injected dependencies to runtime_args, TBD in https://github.com/jina-ai/jina/issues/5396
- [x] implemented Gateway classes uses the port and protocol from runtime args
- [x] support in JAML
- [x] k8s gateway service
- [x] docker compose gateway service
- [x] health checks
* NetworkChecker
* docker and k8s checks
* Flow pod checks
- [ ] ~~apply the same principle for port_monitoring~~ TBD in a separate PR: https://github.com/jina-ai/jina/issues/5352
- [x] fix gateway output: https://github.com/jina-ai/jina/issues/5345
- [ ] pass CI
- [x] Add tests
follow up: https://github.com/jina-ai/jina/issues/5393
closes: https://github.com/jina-ai/jina/issues/5345, https://github.com/jina-ai/jina/issues/5343

### CODE REVIEW

File: jina/orchestrate/flow/base.py [hunk 1]
```diff
@@ -1430,12 +1430,12 @@ def config_gateway(
...
+        :param port: The port for input data to bind to, default is a random port between [49152, 65535]. In the case of an external Executor (`--external` or `external=True`) this can be a list of ports, separated by commas. Then, every resulting address will be considered as one replica of the Executor.
         :param port_monitoring: The port on which the prometheus server is exposed, default is a random port between [49152, 65535]
         :param prefetch: Number of requests fetched from the client before feeding into the first Executor.
 
               Used to control the speed of data input into a Flow. 0 disables prefetch (1000 requests is the default)
-        :param protocol: Communication protocol between server and client.
+        :param protocol: Possible communication protocols between server and client. Choose any list combination of the following choices: ['grpc', 'http', 'websocket'].
```
Comment (JoanFM) [inline jina/orchestrate/flow/base.py:1438]: any list no, depending on your gateway

File: jina/parsers/orchestrate/pod.py [hunk 2]
```diff
@@ -88,16 +88,24 @@ def mixin_pod_parser(parser, pod_type: str = 'worker'):
...
+        'In the case of an external Executor (`--external` or `external=True`) this can be a list of ports, separated by commas. '
+        'Then, every resulting address will be considered as one replica of the Executor.'
+    )
+
+    if pod_type != 'gateway':
+        gp.add_argument(
+            '--port', '--port-in', type=str, default=str(helper.random_port()), help=port_description
+        )
+    else:
+        gp.add_argument(
```
Comment (JoanFM) [inline jina/parsers/orchestrate/pod.py:102]: can u show me a CLI example of how does this work?

File: jina/parsers/orchestrate/runtimes/remote.py [hunk 3]
```diff
@@ -148,6 +148,24 @@ def mixin_gateway_parser(parser):
...
+    """Add the arguments for the protocol to the gateway parser
+
+    :param parser: the parser configure
+    """
+
+    from jina.enums import GatewayProtocolType
+
+    parser.add_argument(
+        '--protocol',
+        nargs='+',
```
Comment (JoanFM) [inline jina/parsers/orchestrate/runtimes/remote.py:161]: we need to have a plural alias

File: jina/parsers/orchestrate/pod.py [hunk 4]
```diff
@@ -88,16 +88,24 @@ def mixin_pod_parser(parser, pod_type: str = 'worker'):
...
+            '--port',
```
Comment (JoanFM) [inline jina/parsers/orchestrate/pod.py:103]: we need a plural alias for gateway

Comment (JoanFM) [reply jina/parsers/orchestrate/pod.py:103]: I believe with an action we can make sure that these are passed as int to the Python objects so that the code is much cleaner afterwards

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/parsers/orchestrate/pod.py (hunk 2 above)
Comment (alaeddine-13) [reply jina/parsers/orchestrate/pod.py:102]: jina gateway --port 12345 12345
or
jina gateway --port 12345

Comment (JoanFM) [reply jina/parsers/orchestrate/pod.py:102]: why didn't we do like this in the past for other cases?

Comment (alaeddine-13) [reply jina/parsers/orchestrate/pod.py:106]: not sure, probably we were just fine with the comma separated list of values format

File: jina/parsers/orchestrate/pod.py (hunk 4 above)
Comment (JoanFM) [reply jina/parsers/orchestrate/pod.py:113]: Add `ports` plural alias

File: jina/parsers/orchestrate/runtimes/remote.py [hunk 5]
```diff
@@ -148,6 +140,24 @@ def mixin_gateway_parser(parser):
...
+def mixin_gateway_protocol_parser(parser):
+    """Add the arguments for the protocol to the gateway parser
+
+    :param parser: the parser configure
+    """
+
+    from jina.enums import GatewayProtocolType
+
+    parser.add_argument(
+        '--protocol',
```
Comment (JoanFM) [inline jina/parsers/orchestrate/runtimes/remote.py:152]: Add `protocols` plural alias

File: jina/parsers/orchestrate/runtimes/remote.py (hunk 3 above)
Comment (JoanFM) [reply jina/parsers/orchestrate/runtimes/remote.py:153]: What happens if I pass a single one?

File: jina/resources/k8s/template/deployment-gateway.yml [hunk 6]
```diff
@@ -30,8 +30,7 @@ spec:
           imagePullPolicy: {pull_policy}
           command: {command}
           args: {args}
-          ports:
-            - containerPort: {port}
+          ports: {ports-section}
```
Comment (JoanFM) [inline jina/resources/k8s/template/deployment-gateway.yml:33]: why this change?

File: jina/resources/k8s/template/service.yml [hunk 7]
```diff
@@ -7,10 +7,6 @@ metadata:
     app: {target}
 spec:
   type: {type}
-  ports:
-    - port: {port}
-      targetPort: {port}
-      name: port
-      protocol: TCP
+  ports: {service-ports-section}
```
Comment (JoanFM) [inline jina/resources/k8s/template/service.yml:10]: why this change?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/parsers/orchestrate/pod.py (hunk 4 above)
Comment (alaeddine-13) [reply jina/parsers/orchestrate/pod.py:113]: done

File: jina/parsers/orchestrate/runtimes/remote.py (hunk 3 above)
Comment (alaeddine-13) [reply jina/parsers/orchestrate/runtimes/remote.py:154]: It will be read as a list of ports with 1 protocol. You can check the behavior here
tests/unit/parsers/peapods/runtimes/test_port_protocol_parser.py

File: jina/parsers/orchestrate/runtimes/remote.py (hunk 5 above)
Comment (alaeddine-13) [reply jina/parsers/orchestrate/runtimes/remote.py:152]: done

File: jina/resources/k8s/template/deployment-gateway.yml (hunk 6 above)
Comment (alaeddine-13) [reply jina/resources/k8s/template/deployment-gateway.yml:33]: we need a ports section with multiple ports here, rather than 1 port element.
This section is filled using a for loop to expose multiple ports
image

File: jina/resources/k8s/template/service.yml (hunk 7 above)
Comment (alaeddine-13) [reply jina/resources/k8s/template/service.yml:10]: for the service to work properly with multiple ports, we need to put of them as items here, not just 1 item

File: tests/integration/multiple_protocol_gateway/gateway/multiprotocol_gateway.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class MultiProtocolGateway(Gateway):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.http_port = self.runtime_args.port[0]
+        self.grpc_port = self.runtime_args.port[1]
+        self.health_servicer = health.HealthServicer(experimental_non_blocking=True)
+
+    async def setup_http_server(self):
```
Comment (JoanFM) [inline tests/integration/multiple_protocol_gateway/gateway/multiprotocol_gateway.py:22]: make them private, to show what is API and what is custom?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/integration/multiple_protocol_gateway/gateway/multiprotocol_gateway.py (hunk 8 above)
Comment (alaeddine-13) [reply tests/integration/multiple_protocol_gateway/gateway/multiprotocol_gateway.py:22]: makes sense

File: tests/k8s/multiprotocol-gateway/multiprotocol_gateway.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        reflection.enable_server_reflection(service_names, self.grpc_server)
+        self.grpc_server.add_insecure_port(f'{__default_host__}:{self.grpc_port}')
+        await self.grpc_server.start()
+
+    async def setup_server(self):
+        await self.setup_http_server()
+        await self.setup_grpc_server()
+
+    async def run_server(self):
+        print('running servers')
```
Comment (JoanFM) [inline tests/k8s/multiprotocol-gateway/multiprotocol_gateway.py:63]: remove print

File: tests/k8s/test_k8s.py [hunk 10]
```diff
@@ -957,8 +957,8 @@ def test_flow_multiple_protocols_gateway(logger, docker_images, tmpdir):
         protocol=['http', 'grpc'],
     )
 
-    dump_path = os.path.join(str(tmpdir), 'k8s-flow-custom-gateway.yml')
-    namespace = 'flow-custom-gateway'.lower()
+    dump_path = os.path.join(str(tmpdir), 'k8s-flow-multiprotocol-gateway.yml')
+    namespace = 'flow-multiprotocol-gateway'.lower()
```
Comment (JoanFM) [inline tests/k8s/test_k8s.py:961]: why do we need these changes?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py [hunk 11]
```diff
@@ -88,19 +88,50 @@ def get_template_yamls(
...
         'protocol': str(protocol).lower() if protocol is not None else '',
-        'volume_path': volumes[0] if volumes is not None else None
+        'volume_path': volumes[0] if volumes is not None else None,
     }
 
     if gpus:
         template_params['device_plugins'] = {'nvidia.com/gpu': gpus}
 
     template_name = 'deployment-executor' if name != 'gateway' else 'deployment-gateway'
+    service_ports_section = '''
```
Comment (JoanFM) [inline jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py:98]: I prefer to  keep this in `yaml`. This gives a single source of truth, and can be more easily changed

Comment (JoanFM) [reply jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py:98]: You can have this in YAML, but then load and edit

File: jina/orchestrate/flow/base.py [hunk 12]
```diff
@@ -2114,22 +2105,31 @@ def _mermaid_to_url(self, mermaid_str: str, img_type: str) -> str:
         return f'https://mermaid.ink/{img_type}/{encoded_str}'
 
     @property
-    def port(self) -> int:
+    def port(self) -> Union[List[int], Optional[int]]:
```
Comment (JoanFM) [inline jina/orchestrate/flow/base.py:2108]: do we have a nice unit test for this property?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/k8s/test_k8s.py (hunk 10 above)
Comment (alaeddine-13) [reply tests/k8s/test_k8s.py:961]: the whole test is named `test_flow_multiple_protocols_gateway` while `test_flow_with_custom_gateway` is unchanged

File: jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py (hunk 11 above)
Comment (alaeddine-13) [reply jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py:98]: will put only this section as a separate file in resources

File: tests/integration/multiple_protocol_gateway/gateway/multiprotocol_gateway.py (hunk 8 above)
Comment (alaeddine-13) [reply tests/integration/multiple_protocol_gateway/gateway/multiprotocol_gateway.py]: done

File: tests/k8s/multiprotocol-gateway/multiprotocol_gateway.py (hunk 9 above)
Comment (alaeddine-13) [reply tests/k8s/multiprotocol-gateway/multiprotocol_gateway.py]: done

File: jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py (hunk 11 above)
Comment (alaeddine-13) [reply jina/orchestrate/deployments/config/k8slib/kubernetes_deployment.py]: done

File: jina/serve/runtimes/gateway/__init__.py [hunk 13]
```diff
@@ -79,6 +78,8 @@ async def async_setup(self):
             uses_metas={},
             runtime_args={  # these are not parsed to the yaml config file but are pass directly during init
                 'name': self.args.name,
+                'port': self.args.port,
```
Comment (samsja) [inline jina/serve/runtimes/gateway/__init__.py:81]: why is it moved to the runtime args now ?

Comment (alaeddine-13) [reply jina/serve/runtimes/gateway/__init__.py:81]: port and protocol are now parameters that the runtime is aware of (used for health checks, k8s service, docker compose expose, ...)
The user no more has control over this parameter and he can't define how this parameter looks like, it is provided by the runtime for him

Comment (samsja) [reply jina/serve/runtimes/gateway/__init__.py:81]: so the user cannot change the port for example ?

Comment (alaeddine-13) [reply jina/serve/runtimes/gateway/__init__.py:81]: He passes a port as input according to our runtime defined format:
```python
Flow().config_gateway(port=12345)
```
or
```python
Flow().config_gateway(port=[12345, 12344])
```
and will read it from runtime_args when he implements the gateway.
He cannot come up with his own format or be able to customize it, like
```python
Flow().config_gateway(port={'http': 12345, 'grpc': 12344)
```

Comment (samsja) [reply jina/serve/runtimes/gateway/__init__.py:81]: okay I see makes sense

File: tests/k8s/test_k8s.py [hunk 14]
```diff
@@ -941,6 +943,69 @@ async def test_flow_with_custom_gateway(logger, docker_images, tmpdir):
...
+        core_client.list_namespaced_pod(
+            namespace=namespace, label_selector='app=gateway'
+        )
+        .items[0]
+        .metadata.name
+    )
+    config_path = os.environ['KUBECONFIG']
+    import portforward
+
+    with portforward.forward(
```
Comment (girishc13) [inline tests/k8s/test_k8s.py:993]: Testing the request using the service is also useful. I don't know if the python kubectl client can do a proxy but on local we can do something like
```
# open a proxy default port at 8081
kubectl proxy &
```
```
curl http://localhost:8081/api/v1/namespaces/${namepsace}/services/gateway:${port}/proxy
```

Comment (samsja) [reply tests/k8s/test_k8s.py:993]: I believe proxy is overkill for that. You can port forward on the service instead of the Deployment.  

`kubectl port-forward svc/NAME port1:port2`

Comment (girishc13) [reply tests/k8s/test_k8s.py]: Anything is fine as long as the service interface is used for the test.

Comment (alaeddine-13) [reply tests/k8s/test_k8s.py]: done

**Review by JoanFM — APPROVED** [review]

**Review by JoanFM — APPROVED** [review]
